src/Shared/Components/ComponentDescriptorSerializer.Deserialize.cs

```diff
@@ -0,0 +1,215 @@
+            RootComponentTypeCache rootComponentTypeCache)
+        {
+            DataProtector = dataProtectionProvider.CreateProtector(DataProtectionProviderPurpose);
+            Logger = logger;
+            RootComponentTypeCache = rootComponentTypeCache;
+        }
+
+        public ILogger<ComponentDescriptorSerializer> Logger { get; }
+
+        public RootComponentTypeCache RootComponentTypeCache { get; }
+
+        public bool TryDeserializeComponentDescriptorCollection(string serializedComponentRecords, out List<ComponentDescriptor> descriptors)
+        {
+            var markers = JsonSerializer.Deserialize<IEnumerable<ComponentMarker>>(serializedComponentRecords, _jsonSerializationOptions);
+            descriptors = new List<ComponentDescriptor>();
+            int lastSequence = -1;
+
+            var previousInstance = new ComponentDescriptorInstance();
+            foreach (var marker in markers)
+            {
+                if (marker.Type != ComponentMarker.ServerMarkerType)
+                {
+                    Log.InvalidMarkerType(Logger, marker.Type);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (marker.Sequence == null)
+                {
+                    Log.MissingMarkerSequence(Logger);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (marker.Descriptor == null)
+                {
+                    Log.MissingMarkerDescriptor(Logger);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                var (descriptor, instance) = DeserializeComponentDescriptor(marker);
+                if (descriptor == null)
+                {
+                    // We failed to deserialize the component descriptor for some reason.
+                    descriptors.Clear();
+                    return false;
+                }
+
+                // We force our client to send the descriptors in order so that we do minimal work.
+                if (lastSequence != -1 && lastSequence != instance.Sequence - 1)
+                {
+                    Log.OutOfSequenceDescriptor(Logger, lastSequence, instance.Sequence);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (lastSequence != -1 &&
+                    !CryptographicOperations.FixedTimeEquals(
+                        MemoryMarshal.AsBytes(previousInstance.InvocationId.AsSpan()),
+                        MemoryMarshal.AsBytes(instance.InvocationId.AsSpan())))
+                {
+                    Log.MismatchedInvocationId(Logger, previousInstance.InvocationId, instance.InvocationId);
+                    return false;
+                }
+
+                // As described below, we build a chain of descriptors to prevent being flooded by
+                // descriptors from a client not behaving properly.
```

Comment:
Well we can't be totally flooded right due to max message size? If that's the concern it seems like we could mitigate it in a few ways.

Reply:
Max message size is the worst case scenario, we can do way better than that, limiting the amount of rendered components to those generated as part of an HTTP response by the app. Which will be a way lower number, more predictable and app dependent, so the app decides on a per request basis how many components are acceptable to render. with MAX message size being the only limitation in the worst case scenario.